From 2de748a3e787140c528064a1ceae3c98bc651aaf Mon Sep 17 00:00:00 2001
From: Parth Gajjar <parth.gajjar@amd.com>
Date: Mon, 3 Mar 2025 04:07:16 -0800
Subject: [PATCH 3/4] Making gpu driver compatible with 6.12 kernel

Remove returns void after 6.11 kernel instead of int

Fix compilation issue for 6.12 kernel:

Fixed MIN MAX redefinition
Fix FMODE_UNSIGNED_OFFSET compilation error
Fix dma_set_max_seg_size returns NULL
Update no_llseek function to noop_llseek

Signed-off-by: Vishal Sagar <vishal.sagar@amd.com>
Signed-off-by: Parth Gajjar <parth.gajjar@amd.com>
---
 .../memory_group_manager/memory_group_manager.c |  8 ++++++++
 .../protected_memory_allocator.c                | 12 ++++++++++++
 .../gpu/arm/midgard/device/mali_kbase_device.c  |  8 ++++++++
 drivers/gpu/arm/midgard/mali_kbase_core_linux.c | 17 +++++++++++++++++
 drivers/gpu/arm/midgard/mali_kbase_kinstr_jm.c  |  4 ++++
 drivers/gpu/arm/midgard/mali_malisw.h           |  3 +++
 .../platform/devicetree/mali_kbase_runtime_pm.c |  4 +++-
 7 files changed, 55 insertions(+), 1 deletion(-)

diff --git a/drivers/base/arm/memory_group_manager/memory_group_manager.c b/drivers/base/arm/memory_group_manager/memory_group_manager.c
index da4a0c3..3976e8b 100644
--- a/drivers/base/arm/memory_group_manager/memory_group_manager.c
+++ b/drivers/base/arm/memory_group_manager/memory_group_manager.c
@@ -446,7 +446,11 @@ static int memory_group_manager_probe(struct platform_device *pdev)
 	return 0;
 }
 
+#if (KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE)
+static void memory_group_manager_remove(struct platform_device *pdev)
+#else
 static int memory_group_manager_remove(struct platform_device *pdev)
+#endif
 {
 	struct memory_group_manager_device *mgm_dev = platform_get_drvdata(pdev);
 	struct mgm_groups *mgm_data = mgm_dev->data;
@@ -458,7 +462,11 @@ static int memory_group_manager_remove(struct platform_device *pdev)
 
 	dev_info(&pdev->dev, "Memory group manager removed successfully\n");
 
+#if (KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE)
+	return;
+#else
 	return 0;
+#endif
 }
 
 static const struct of_device_id memory_group_manager_dt_ids[] = {
diff --git a/drivers/base/arm/protected_memory_allocator/protected_memory_allocator.c b/drivers/base/arm/protected_memory_allocator/protected_memory_allocator.c
index b27eed9..2a0f99a 100644
--- a/drivers/base/arm/protected_memory_allocator/protected_memory_allocator.c
+++ b/drivers/base/arm/protected_memory_allocator/protected_memory_allocator.c
@@ -495,14 +495,22 @@ static int protected_memory_allocator_probe(struct platform_device *pdev)
 	return 0;
 }
 
+#if (KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE)
+static void protected_memory_allocator_remove(struct platform_device *pdev)
+#else
 static int protected_memory_allocator_remove(struct platform_device *pdev)
+#endif
 {
 	struct protected_memory_allocator_device *pma_dev = platform_get_drvdata(pdev);
 	struct simple_pma_device *epma_dev;
 	struct device *dev;
 
 	if (!pma_dev)
+#if (KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE)
+		return;
+#else
 		return -EINVAL;
+#endif
 
 	epma_dev = container_of(pma_dev, struct simple_pma_device, pma_dev);
 	dev = epma_dev->dev;
@@ -518,7 +526,11 @@ static int protected_memory_allocator_remove(struct platform_device *pdev)
 
 	dev_info(&pdev->dev, "Protected memory allocator removed successfully\n");
 
+#if (KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE)
+	return;
+#else
 	return 0;
+#endif
 }
 
 static const struct of_device_id protected_memory_allocator_dt_ids[] = {
diff --git a/drivers/gpu/arm/midgard/device/mali_kbase_device.c b/drivers/gpu/arm/midgard/device/mali_kbase_device.c
index 670b2c2..f3b0f18 100644
--- a/drivers/gpu/arm/midgard/device/mali_kbase_device.c
+++ b/drivers/gpu/arm/midgard/device/mali_kbase_device.c
@@ -31,6 +31,7 @@
 #include <linux/of_platform.h>
 #include <linux/types.h>
 #include <linux/oom.h>
+#include <linux/version.h>
 
 #include <mali_kbase.h>
 #include <mali_kbase_defs.h>
@@ -308,9 +309,16 @@ int kbase_device_misc_init(struct kbase_device *const kbdev)
 
 	/* There is no limit for Mali, so set to max. */
 	if (kbdev->dev->dma_parms)
+#if (KERNEL_VERSION(6, 12, 0) > LINUX_VERSION_CODE)
 		err = dma_set_max_seg_size(kbdev->dev, UINT_MAX);
+#else
+		dma_set_max_seg_size(kbdev->dev, UINT_MAX);
+#endif
+
+#if (KERNEL_VERSION(6, 12, 0) > LINUX_VERSION_CODE)
 	if (err)
 		goto dma_set_mask_failed;
+#endif
 
 	kbdev->nr_hw_address_spaces = (s8)kbdev->gpu_props.num_address_spaces;
 
diff --git a/drivers/gpu/arm/midgard/mali_kbase_core_linux.c b/drivers/gpu/arm/midgard/mali_kbase_core_linux.c
index cb113ed..830351a 100644
--- a/drivers/gpu/arm/midgard/mali_kbase_core_linux.c
+++ b/drivers/gpu/arm/midgard/mali_kbase_core_linux.c
@@ -726,7 +726,9 @@ static int kbase_open(struct inode *inode, struct file *filp)
 	}
 
 	filp->private_data = kfile;
+#if (KERNEL_VERSION(6, 12, 0) > LINUX_VERSION_CODE)
 	filp->f_mode |= FMODE_UNSIGNED_OFFSET;
+#endif
 
 	return 0;
 
@@ -2188,6 +2190,9 @@ static const struct file_operations kbase_fops = {
 	.mmap = kbase_mmap,
 	.check_flags = kbase_check_flags,
 	.get_unmapped_area = kbase_get_unmapped_area,
+#if (KERNEL_VERSION(6, 12, 0) <= LINUX_VERSION_CODE)
+	.fop_flags = FOP_UNSIGNED_OFFSET,
+#endif
 };
 
 /**
@@ -5705,18 +5710,30 @@ void kbase_sysfs_term(struct kbase_device *kbdev)
 	put_device(kbdev->dev);
 }
 
+#if (KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE)
+static void kbase_platform_device_remove(struct platform_device *pdev)
+#else
 static int kbase_platform_device_remove(struct platform_device *pdev)
+#endif
 {
 	struct kbase_device *kbdev = to_kbase_device(&pdev->dev);
 
 	if (!kbdev)
+#if (KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE)
+		return;
+#else
 		return -ENODEV;
+#endif
 
 	kbase_device_term(kbdev);
 	dev_set_drvdata(kbdev->dev, NULL);
 	kbase_device_free(kbdev);
 
+#if (KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE)
+	return;
+#else
 	return 0;
+#endif
 }
 
 void kbase_backend_devfreq_term(struct kbase_device *kbdev)
diff --git a/drivers/gpu/arm/midgard/mali_kbase_kinstr_jm.c b/drivers/gpu/arm/midgard/mali_kbase_kinstr_jm.c
index 0bf1450..0dd82f7 100644
--- a/drivers/gpu/arm/midgard/mali_kbase_kinstr_jm.c
+++ b/drivers/gpu/arm/midgard/mali_kbase_kinstr_jm.c
@@ -651,7 +651,11 @@ static __poll_t reader_poll(struct file *const file, struct poll_table_struct *c
 
 /* The file operations virtual function table */
 static const struct file_operations file_operations = { .owner = THIS_MODULE,
+#if (KERNEL_VERSION(6, 12, 0) > LINUX_VERSION_CODE)
 							.llseek = no_llseek,
+#else
+							.llseek = noop_llseek,
+#endif
 							.read = reader_read,
 							.poll = reader_poll,
 							.release = reader_release };
diff --git a/drivers/gpu/arm/midgard/mali_malisw.h b/drivers/gpu/arm/midgard/mali_malisw.h
index f34ba80..45f2e20 100644
--- a/drivers/gpu/arm/midgard/mali_malisw.h
+++ b/drivers/gpu/arm/midgard/mali_malisw.h
@@ -22,10 +22,12 @@
 /*
  * Kernel-wide include for common macros and types.
  */
+#include <linux/version.h>
 
 #ifndef _MALISW_H_
 #define _MALISW_H_
 
+#if (KERNEL_VERSION(6, 11, 0) > LINUX_VERSION_CODE)
 /**
  * MIN - Return the lesser of two values.
  * @x: value1
@@ -49,6 +51,7 @@
  * instead.
  */
 #define MAX(x, y) ((x) < (y) ? (y) : (x))
+#endif
 
 /**
  * CSTD_UNUSED - Function-like macro for suppressing unused variable warnings.
diff --git a/drivers/gpu/arm/midgard/platform/devicetree/mali_kbase_runtime_pm.c b/drivers/gpu/arm/midgard/platform/devicetree/mali_kbase_runtime_pm.c
index 9e3f789..9dfd2e9 100644
--- a/drivers/gpu/arm/midgard/platform/devicetree/mali_kbase_runtime_pm.c
+++ b/drivers/gpu/arm/midgard/platform/devicetree/mali_kbase_runtime_pm.c
@@ -77,7 +77,9 @@ static void disable_gpu_power_control(struct kbase_device *kbdev)
 static int pm_callback_power_on(struct kbase_device *kbdev)
 {
 	int ret = 1; /* Assume GPU has been powered off */
-	int error;
+#if MALI_USE_CSF | KBASE_PM_RUNTIME
+	int error = 0;
+#endif
 	unsigned long flags;
 
 	dev_dbg(kbdev->dev, "%s %pK\n", __func__, (void *)kbdev->dev->pm_domain);
-- 
2.34.1

